rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RemoteFetchProcessor.java

```diff
@@ -0,0 +1,114 @@
+ * limitations under the License. See accompanying LICENSE file.
+ */
+package com.qubole.rubix.bookkeeper;
+
+import com.google.common.util.concurrent.AbstractScheduledService;
+import com.qubole.rubix.core.FileDownloadRequestChain;
+import com.qubole.rubix.spi.CacheConfig;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.ExecutionException;
+
+public class RemoteFetchProcessor extends AbstractScheduledService
+{
+  private Configuration conf;
+  private Queue<FetchRequest> processQueue = null;
+  private FileDownloader downloader = null;
+
+  int processThreadInitalDelay;
+  int processThreadInterval;
+  long requestProcessDelay;
+
+  private static final Log log = LogFactory.getLog(RemoteFetchProcessor.class);
+
+  public RemoteFetchProcessor(Configuration conf)
+  {
+    this.conf = conf;
+    this.processQueue = new ConcurrentLinkedQueue<FetchRequest>();
+    this.downloader = new FileDownloader(conf);
+
+    this.processThreadInitalDelay = CacheConfig.getProcessThreadInitialDelayInMs(conf);
+    this.processThreadInterval = CacheConfig.getProcessThreadIntervalInMs(conf);
+    this.requestProcessDelay = CacheConfig.getRemoteFetchProcessIntervalInMS(conf);
+  }
+
+  public void addToProcessQueue(String remotePath, long offset, int length, long fileSize, long lastModified)
+  {
+    long requestedTime = System.currentTimeMillis();
+    FetchRequest request = new FetchRequest(remotePath, offset, length, fileSize, lastModified, requestedTime);
+    processQueue.add(request);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception
+  {
+    long currentTime = System.currentTimeMillis();
+
+    processRequest(currentTime);
+  }
+
+  protected void processRequest(long currentTime) throws IOException, InterruptedException, ExecutionException
+  {
+    ConcurrentMap<String, DownloadRequestContext> contextMap = mergeRequests(currentTime);
+    processRemoteFetchRequest(contextMap);
+
+    // After every iteration we are clearing the map
+    contextMap.clear();
+  }
+
+  protected ConcurrentMap<String, DownloadRequestContext> mergeRequests(long currentTime)
+  {
+    // Till the queue is not empty or there are no more requests which came in before the configured delay time
+    // we are going to collect the requests and process them
+    ConcurrentMap<String, DownloadRequestContext> contextMap = new ConcurrentHashMap<String, DownloadRequestContext>();
+    while (!processQueue.isEmpty()) {
+      FetchRequest request = processQueue.peek();
+      if (currentTime - request.getRequestedTime() < this.requestProcessDelay) {
```

Comment:
I think this can be further optimized. Do we really want to keep the request on the queue even though it could be merged to the existing DownloadRequestContext? We could rather make an exception in such a case and pick this FetchRequest if it could be merged in one of the DownloadRequess

Reply:
Can we take this an enhancement and add that logic as a separate PR ?

Reply:
we can take it as enhancement but do open a jira to track it